src/entities/addLiquidity/addLiquidityCowAmm/index.ts

```diff
@@ -0,0 +1,135 @@
+import { HumanAmount } from '@/data';
+import { Token } from '@/entities/token';
+import { TokenAmount } from '@/entities/tokenAmount';
+import { PoolState, PoolStateWithBalances } from '@/entities/types';
+import {
+    calculateProportionalAmounts,
+    getSortedTokens,
+    getValue,
+} from '@/entities/utils';
+import { CHAINS } from '@/utils';
+
+import { getAmountsCall } from '../helpers';
+import {
+    AddLiquidityBase,
+    AddLiquidityBaseBuildCallInput,
+    AddLiquidityBaseQueryOutput,
+    AddLiquidityBuildCallOutput,
+    AddLiquidityKind,
+    AddLiquidityProportionalInput,
+} from '../types';
+import {
+    getPoolTokenBalanceCowAmm,
+    getTotalSupplyCowAmm,
+} from '../../utils/cowAmmHelpers';
+
+export class AddLiquidityCowAmm implements AddLiquidityBase {
+    async query(
+        input: AddLiquidityProportionalInput,
+        poolState: PoolState,
+    ): Promise<AddLiquidityBaseQueryOutput> {
+        const sortedTokens = getSortedTokens(poolState.tokens, input.chainId);
+
+        const client = createPublicClient({
+            transport: http(input.rpcUrl),
+            chain: CHAINS[input.chainId],
+        });
+
+        const balances: bigint[] = [];
+        for (const token of sortedTokens) {
+            balances.push(
+                await getPoolTokenBalanceCowAmm(
+                    poolState.id,
+                    token.address,
+                    client,
+                ),
+            );
+        }
+
+        const totalShares = await getTotalSupplyCowAmm(
+            poolState.address,
+            client,
+        );
+
+        const poolStateWithBalances: PoolStateWithBalances = {
+            ...poolState,
+            tokens: sortedTokens.map((token, i) => ({
+                address: token.address,
+                decimals: token.decimals,
+                index: i,
+                balance: formatUnits(
+                    balances[i],
+                    token.decimals,
+                ) as HumanAmount,
+            })),
+            totalShares: formatEther(totalShares) as HumanAmount,
+        };
+
+        const { tokenAmounts, bptAmount } = calculateProportionalAmounts(
+            poolStateWithBalances,
+            input.bptOut,
+        );
+        const bptOut = TokenAmount.fromRawAmount(
+            new Token(input.chainId, bptAmount.address, bptAmount.decimals),
+            bptAmount.rawAmount,
+        );
+        const amountsIn = tokenAmounts.map((amountIn) =>
+            TokenAmount.fromRawAmount(
+                new Token(input.chainId, amountIn.address, amountIn.decimals),
+                amountIn.rawAmount,
+            ),
+        );
+        const tokenInIndex = undefined;
+
+        const output: AddLiquidityBaseQueryOutput = {
+            poolType: poolStateWithBalances.type,
+            poolId: poolStateWithBalances.id,
+            addLiquidityKind: input.kind,
+            bptOut,
+            amountsIn,
+            tokenInIndex,
```

Comment:
We don't have to explicitly return this.

Reply:
This is actually one of the side effects of reusing v2/v3 flow for CowAmm. 
I agree that `tokenInIndex` is indeed not relevant for CowAmm, but AddLiquidityProportional for v2/v3 do return this explicitly and I believe we should keep the same pattern.
This gets highlighted in the test helper that expects this attribute to be set [here](https://github.com/balancer/b-sdk/blob/925a4d8928f4bfa638f9f23b090f6f4d95d73667/test/lib/utils/addLiquidityHelper.ts#L335).
I'd have to add an if clause there and skip this check specifically for CowAmm pools, which I don't think make much sense.
I'd lean towards leaving as is, but let me know if you still prefer to remove it and I'll do it over here 😉

Reply:
Thanks for the info, happy with that.